File: src/main/java/at/flauschigesalex/minecraftLibrary/bukkit/PluginCommand.java
package at.flauschigesalex.minecraftLibrary.bukkit;

import at.flauschigesalex.minecraftLibrary.FlauschigeMinecraftLibrary;
import lombok.Getter;
import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.command.Command;
import org.bukkit.command.CommandMap;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;
import org.jetbrains.annotations.Unmodifiable;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static at.flauschigesalex.minecraftLibrary.FlauschigeMinecraftLibrary.getLibrary;

@SuppressWarnings({"unused", "deprecation"})
@Getter
public abstract class PluginCommand extends Command {

    private String pluginPrefix;
    private final String command;

    protected PluginCommand(@NotNull String command) {
        this(command, "");
    }

    protected PluginCommand(@NotNull String command, @NotNull String description) {
        this(command, "", "/" + command);
    }

    protected PluginCommand(@NotNull String command, @NotNull String description, @NotNull String usage) {
        this(command, "", "/" + command, new ArrayList<>());
    }

    protected PluginCommand(@NotNull String command, @NotNull String description, @NotNull String usage, @NotNull ArrayList<String> aliases) {
        super(command, description, usage, aliases);
        final String pluginName = FlauschigeMinecraftLibrary.getPluginName();
        this.pluginPrefix = pluginName == null ? "flauschigesalex" : pluginName.toLowerCase();

        this.command = command;
    }

    public final boolean register(final @NotNull CommandMap commandMap) {
        return super.register(commandMap);
    }

    public final boolean unregister(final @NotNull CommandMap commandMap) {
        return super.unregister(commandMap);
    }

    public final boolean isRegistered() {
        return super.isRegistered();
    }

    public final @NotNull String getName() {
        return super.getName();
    }

    public final @Deprecated boolean setName(final @NotNull String name) {
        return super.setName(name);
    }

    public final @Nullable String getPermission() {
        return super.getPermission();
    }

    public final void setPermission(final @Nullable String permission) {
        super.setPermission(permission);
    }

    /**
     * @see #permissible(CommandSender)
     * @deprecated
     */
    public final boolean testPermission(final @NotNull CommandSender permissible) {
        return this.permissible(permissible);
    }

    /**
     * @deprecated
     */
    public final boolean testPermissionSilent(final @NotNull CommandSender target) {
        return super.testPermissionSilent(target);
    }

    public final @NotNull String getLabel() {
        getLibrary();
        if (FlauschigeMinecraftLibrary.getPluginName() != null)
            return FlauschigeMinecraftLibrary.getPluginName();
        
        return super.getLabel();
    }

    public final boolean setLabel(final @NotNull String name) {
        return super.setLabel(name);
    }

    /**
     * @see #permissionMessage()
     * @deprecated
     */
    @SuppressWarnings("deprecation")
    public final @Nullable String getPermissionMessage() {
        return super.getPermissionMessage();
    }

    /**
     * @see #permissionMessage(Component)
     * @deprecated
     */
    @SuppressWarnings("deprecation")
    public final @NotNull Command setPermissionMessage(final @Nullable String permissionMessage) {
        return super.setPermissionMessage(permissionMessage);
    }

    public final @NotNull String getDescription() {
        return super.getDescription();
    }

    public final @NotNull Command setDescription(final @NotNull String description) {
        return super.setDescription(description);
    }

    public final @NotNull String getUsage() {
        return super.getUsage();
    }

    /**
     * @see #PluginCommand(String, String, String)
     */
    public final @NotNull Command setUsage(final @NotNull String usage) {
        return super.setUsage(usage);
    }

    public final @NotNull List<String> getAliases() {
        return super.getAliases();
    }

    /**
     * @see #PluginCommand(String, String, String, ArrayList)
     */
    public final @NotNull Command setAliases(final @NotNull List<String> aliases) {
        return super.setAliases(aliases);
    }

    public final @Nullable Component permissionMessage() {
        return super.permissionMessage();
    }

    public final void permissionMessagefinal(final @Nullable Component permissionMessage) {
        super.permissionMessage(permissionMessage);
    }

    public final @Deprecated boolean execute(@NotNull CommandSender commandSender, @NotNull String s, @NotNull String[] strings) {
        this.executeCommand(commandSender, s, strings);
        return true;
    }
    protected abstract void executeCommand(final @NotNull CommandSender sender, final @NotNull String fullCommand, final @NotNull String[] args);

    protected Set<TabComplete> tabCompletes(final CommandSender sender) {
        return TabComplete.onlinePlayers(sender instanceof Player player ? player : null);
    }

    public final @Deprecated @NotNull List<String> tabComplete(final @NotNull CommandSender sender, final @NotNull String alias, final @NotNull String[] args) throws IllegalArgumentException {
        return List.of();
    }

    public final @Deprecated @NotNull List<String> tabComplete(final @NotNull CommandSender sender, final @NotNull String alias, final @NotNull String[] args, final @Nullable Location location) throws IllegalArgumentException {
        return tabCompletes(sender).stream()
                .filter(complete -> {
                    if (complete.arg == null)
                        return true;

                    return complete.arg == args.length - 1;
                })
                .filter(complete -> {
                    final int arg = args.length-1;
                    if (arg < 0)
                        return true;

                    if (args[arg].isBlank())
                        return true;

                    return complete.completable.toLowerCase().startsWith(args[arg].toLowerCase());
                })
                .filter(complete -> {
                    if (complete.location == null || complete.location.getWorld() == null)
                        return true;

                    if (location == null || location.getWorld() == null)
                        return false;

                    if (complete.maxDistance < 0)
                        return complete.location.getWorld().equals(location.getWorld());

                    return complete.location.distance(location) <= complete.maxDistance;
                })
                .map(complete -> complete.completable).toList();
    }

    public boolean permissible(final @NotNull CommandSender permissible) {
        return super.testPermission(permissible);
    }

    public PluginCommand setPluginPrefix(final @NotNull String pluginPrefix) {
        this.pluginPrefix = pluginPrefix;
        return this;
    }

    public static final class TabComplete {
        public static Set<TabComplete> onlinePlayers() {
            return onlinePlayers(null);
        }

        public static Set<TabComplete> onlinePlayers(final @Nullable Player player) {
            return new HashSet<>(Bukkit.getOnlinePlayers().stream()
                    .filter(onlinePlayer -> {
                        if (player == null)
                            return true;

                        return player.canSee(onlinePlayer);
                    })
                    .map(onlinePlayer -> new TabComplete(onlinePlayer.getName())).toList());
        }

        private final Integer arg;
        private final String completable;

        private final Location location;
        private final double maxDistance;

        public TabComplete(final @NotNull String completable) {
            this(null, completable, null);
        }
        public TabComplete(final @Nullable Integer arg, final @NotNull String completable) {
            this(arg, completable, null);
        }
        public TabComplete(final @Nullable Integer arg, final @NotNull String completable, final @Nullable World requiredWorld) {
            this(arg, completable, new Location(requiredWorld, 0, 0, 0), -1);
        }
        public TabComplete(final @Nullable Integer arg, final @NotNull String completable, final @Nullable Location location, final @Range(from = 0, to = Long.MAX_VALUE) double maxDistance) {
            this.arg = arg;
            this.completable = completable;

            this.location = location;
            this.maxDistance = maxDistance;
        }

        static @Unmodifiable List<TabComplete> players(final int arg) {
            return Bukkit.getOnlinePlayers().stream().map(player -> new TabComplete(arg, player.getName())).toList();
        }

        public boolean equals(Object obj) {
            if (obj instanceof TabComplete complete)
                return arg == complete.arg && completable.equalsIgnoreCase(complete.completable);

            return false;
        }
    }
}

File: src/main/java/at/flauschigesalex/minecraftLibrary/bukkit/PluginCommand.kt
@file:Suppress("DeprecatedCallableAddReplaceWith", "MemberVisibilityCanBePrivate", "UNUSED_PARAMETER", "CanBeParameter")

package at.flauschigesalex.minecraftLibrary.bukkit

import at.flauschigesalex.minecraftLibrary.FlauschigeMinecraftLibrary.Companion.library
import at.flauschigesalex.minecraftLibrary.FlauschigeMinecraftLibrary.Companion.pluginName
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.World
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.jetbrains.annotations.Range
import org.jetbrains.annotations.Unmodifiable
import java.util.*

@Suppress("unused", "DEPRECATION")
abstract class PluginCommand protected constructor(val command: String, description: String, usage: String, aliases: ArrayList<String?>)
    : Command(command, description, usage, aliases) {

    var pluginPrefix: String
        private set

    protected constructor(command: String, description: String = "") : this(command, "", "/$command")

    protected constructor(command: String, description: String, usage: String) : this(
        command, "",
        "/$command", ArrayList<String?>()
    )

    init {
        val pluginName = pluginName
        this.pluginPrefix = pluginName?.lowercase(Locale.getDefault()) ?: "flauschigesalex"
    }

    @Deprecated("", ReplaceWith("this.command")) override fun setName(name: String): Boolean {
        return super.setName(name)
    }

    @Deprecated("", ReplaceWith("this.permissible(permissible)"))
    override fun testPermission(permissible: CommandSender): Boolean {
        return this.permissible(permissible)
    }

    @Deprecated("", ReplaceWith("this.permissible(permissible)"))
    override fun testPermissionSilent(target: CommandSender): Boolean {
        return super.testPermissionSilent(target)
    }

    override fun getLabel(): String {
        library
        if (pluginName != null) return pluginName!!

        return super.getLabel()
    }

    @Deprecated("") override fun getPermissionMessage(): String? {
        return super.getPermissionMessage()
    }

    @Deprecated("") override fun setPermissionMessage(permissionMessage: String?): Command {
        return super.setPermissionMessage(permissionMessage)
    }

    override fun setAliases(aliases: List<String>): Command {
        return super.setAliases(aliases)
    }

    @Deprecated("") override fun execute(commandSender: CommandSender, s: String, strings: Array<String>): Boolean {
        this.executeCommand(commandSender, s, strings)
        return true
    }

    protected abstract fun executeCommand(sender: CommandSender, fullCommand: String, args: Array<String>)

    protected fun tabCompletes(sender: CommandSender?): Set<TabComplete> {
        return TabComplete.onlinePlayers(if (sender is Player) sender else null)
    }

    @Deprecated("") @Throws(IllegalArgumentException::class) override fun tabComplete(
        sender: CommandSender,
        alias: String,
        args: Array<String>
    ): List<String> {
        return listOf()
    }

    @Deprecated("") @Throws(IllegalArgumentException::class)
    override fun tabComplete(sender: CommandSender, alias: String, args: Array<String>, location: Location?): List<String> {
        return tabCompletes(sender).stream()
            .filter { complete: TabComplete ->
                if (complete.arg == null) return@filter true
                complete.arg == args.size - 1
            }
            .filter { complete: TabComplete ->
                val arg = args.size - 1
                if (arg < 0) return@filter true

                if (args[arg].isBlank()) return@filter true
                complete.completable.lowercase(Locale.getDefault()).startsWith(args[arg].lowercase(Locale.getDefault()))
            }
            .filter { complete: TabComplete ->
                if (complete.location == null || complete.location.world == null) return@filter true
                if (location == null || location.world == null) return@filter false

                if (complete.maxDistance < 0) return@filter complete.location.world == location.world
                complete.location.distance(location) <= complete.maxDistance
            }
            .map { complete: TabComplete -> complete.completable }.toList()
    }

    fun permissible(permissible: CommandSender): Boolean {
        return super.testPermission(permissible)
    }

    fun setPluginPrefix(pluginPrefix: String): PluginCommand {
        this.pluginPrefix = pluginPrefix
        return this
    }

    class TabComplete(val arg: Int?, val completable: String, val location: Location?,
                      val maxDistance: @Range(from = 0, to = Long.MAX_VALUE) Double) {

        companion object {
            fun onlinePlayers(player: Player? = null): Set<TabComplete> {
                return HashSet(Bukkit.getOnlinePlayers().stream()
                    .filter { onlinePlayer: Player? ->
                        if (player == null) return@filter true
                        player.canSee(onlinePlayer!!)
                    }
                    .map { onlinePlayer: Player? -> TabComplete(onlinePlayer!!.name) }
                    .toList())
            }

            fun players(arg: Int): @Unmodifiable MutableList<TabComplete> {
                return Bukkit.getOnlinePlayers().stream()
                    .map { player: Player? -> TabComplete(arg, player!!.name) }
                    .toList()
            }
        }

        constructor(completable: String) : this(null, completable, null)

        constructor(arg: Int?, completable: String, requiredWorld: World? = null)
                : this(arg, completable, Location(requiredWorld, 0.0, 0.0, 0.0), -1.0)

        override fun equals(other: Any?): Boolean {
            if (other is TabComplete) return arg == other.arg && completable.equals(other.completable, ignoreCase = true)

            return false
        }

        override fun hashCode(): Int {
            return javaClass.hashCode()
        }
    }
}
